src/ck3/data/characters.rs

```diff
@@ -301,125 +301,270 @@ impl Character {
+                        }
+                        Some((AddSpouse, value))
+                    }
+                    "add_same_sex_spouse" => {
+                        data.characters.verify_exists_gender(value, gender);
+                        if data.item_exists(Item::Character, value.as_str()) {
+                            data.characters.verify_alive(value, date);
+                        }
+                        Some((AddSpouse, value))
+                    }
+                    "add_concubine" => {
+                        data.characters.verify_exists_gender(value, gender.flip());
+                        if data.item_exists(Item::Character, value.as_str()) {
+                            data.characters.verify_alive(value, date);
+                        }
+                        None
+                    }
+                    "remove_spouse" => Some((RemoveSpouse, value)),
+                    "dynasty" => {
+                        data.verify_exists(Item::Dynasty, value);
+                        None
+                    }
+                    "dynasty_house" => {
+                        data.verify_exists(Item::House, value);
+                        None
+                    }
+                    "give_nickname" => {
+                        data.verify_exists(Item::Nickname, value);
+                        None
+                    }
+                    "add_prestige" | "add_piety" | "add_gold" => {
+                        value.expect_precise_number();
+                        None
+                    }
+                    "give_council_position" => {
+                        data.verify_exists(Item::CouncilPosition, value);
+                        Some((GiveCouncilPosition, key))
+                    }
+                    _ => {
+                        let msg = format!("unknown key `{key}` in assignment");
+                        err(ErrorKey::Validation).msg(msg).loc(key).push();
+                        None
+                    }
                 }
             }
-        }
-
-        vd.field_item("culture", Item::Culture);
-        if let Some(token) = vd.field_value("set_culture") {
-            validate_prefix_reference_token(token, data, "culture");
-        }
-
-        vd.multi_field_item("trait", Item::Trait);
-        vd.multi_field_item("add_trait", Item::Trait);
-        vd.multi_field_item("remove_trait", Item::Trait);
+            BV::Block(block) => match key.as_str() {
+                "death" => {
+                    let mut vd = Validator::new(block, data);
+                    vd.req_field("death_reason");
+                    vd.field_item("death_reason", Item::DeathReason);
+                    vd.field_item("killer", Item::Character);
+                    Some((Death, key))
+                }
+                "effect" => {
+                    validate_effect(block, data, sc, Tooltipped::No);
+                    None
+                }
+                _ => {
```

Comment:
Unfortunately you can't replace `validate_effect_internal` with this. Ck3 allows arbitrary effects directly in the history, without an `effect` block.

This also means you can't skip the use of a Validator that marks the non-effect fields as "known" before calling `validate_effect_internal`

Reply:
Okay. I didn’t see any loose effect in the vanilla files. Just for reference, this function is validating each line individually rather than as a block though. So the highlighted line is actually evaluating the effect block.

Reply:
My comment was attached to the `_ => {` line :) Which starts the rejection of all unknown effects. It's a pity that github only shows the lines above and not below